cmd/crank/beta/validate/validate.go

```diff
@@ -17,102 +17,132 @@ limitations under the License.
 				Version: ver.Name,
 				Kind:    internal.Spec.Names.Kind,
 			}
 
-			// Version specific validation rules
-			if ver.Schema != nil && ver.Schema.OpenAPIV3Schema != nil {
-				sv, _, err = validation.NewSchemaValidator(ver.Schema.OpenAPIV3Schema)
-				if err != nil {
-					return nil, err
-				}
+			var s *ext.JSONSchemaProps
+			switch {
+			case internal.Spec.Validation != nil:
+				s = internal.Spec.Validation.OpenAPIV3Schema
+			case ver.Schema != nil && ver.Schema.OpenAPIV3Schema != nil:
+				s = ver.Schema.OpenAPIV3Schema
+			default:
+				// TODO log a warning here, it should never happen
+				continue
+			}
 
-				validators[gvk] = append(validators[gvk], sv)
+			sv, _, err = validation.NewSchemaValidator(s)
+			if err != nil {
+				return nil, nil, err
 			}
 
-			// Top level validation rules
-			if internal.Spec.Validation != nil {
-				sv, _, err = validation.NewSchemaValidator(internal.Spec.Validation.OpenAPIV3Schema)
-				if err != nil {
-					return nil, err
-				}
+			validators[gvk] = append(validators[gvk], &sv)
 
-				validators[gvk] = append(validators[gvk], sv)
+			structural, err := schema.NewStructural(s)
+			if err != nil {
+				return nil, nil, err
 			}
+
+			structurals[gvk] = structural
 		}
 	}
 
-	return validators, nil
+	return validators, structurals, nil
 }
 
 // SchemaValidation validates the resources against the given CRDs
 func SchemaValidation(resources []*unstructured.Unstructured, crds []*extv1.CustomResourceDefinition, skipSuccessLogs bool) error {
-	schemaValidators, err := newValidators(crds)
+	schemaValidators, structurals, err := newValidatorsAndStructurals(crds)
 	if err != nil {
 		return errors.Wrap(err, "cannot create schema validators")
 	}
 
-	failure, warning := 0, 0
+	failure, missingSchemas := 0, 0
 
 	for i, r := range resources {
-		resourceValidators, ok := schemaValidators[r.GetObjectKind().GroupVersionKind()]
+		gvk := r.GetObjectKind().GroupVersionKind()
+		sv, ok := schemaValidators[gvk]
 		if !ok {
-			warning++
+			missingSchemas++
 			fmt.Println("[!] could not find CRD/XRD for: " + r.GroupVersionKind().String())
 			continue
 		}
 
 		rf := 0
-		for _, v := range resourceValidators {
-			re := v.Validate(&resources[i])
-			for _, e := range re.Errors {
+
+		for _, v := range sv {
+			re := validation.ValidateCustomResource(nil, r, *v)
+			for _, e := range re {
 				rf++
-				fmt.Printf("[x] validation error %s, %s : %s\n", r.GroupVersionKind().String(), r.GetAnnotations()[composite.AnnotationKeyCompositionResourceName], e.Error())
+				fmt.Printf("[x] schema validation error %s, %s : %s\n", r.GroupVersionKind().String(), getResourceName(r), e.Error())
 			}
-		}
 
-		if rf == 0 && !skipSuccessLogs {
-			fmt.Printf("[✓] %s, %s validated successfully\n", r.GroupVersionKind().String(), r.GetAnnotations()[composite.AnnotationKeyCompositionResourceName])
-		} else {
-			failure++
+			s := structurals[gvk] // if we have a schema validator, we should also have a structural
```

Comment:
feels like we could merge the two maps by having a dedicated struct wrapping both the schemaValidator and the then CEL validator